prov/sm2/src/sm2.h

```diff
@@ -107,7 +108,10 @@ enum {
  * 	user_data - the message
  */
 struct sm2_xfer_hdr {
-	volatile long int next;
+	union {
+		volatile long int next;
+		struct sm2_ep *ep;
+	};
```

Comment:
Hm.  The fact that we have to hijack a field in the xfer_hdr for use in unexpected messages strikes me as strange.  This particular field will almost certainly make us segfault if we get it wrong.

I wish we had a few extra bytes in the fi_peer_rx_entry to use for this.  If not, I wonder if it would be cleaner to dedicate a field in xfer_hdr to this for clarity.

In any case, the thing that strikes me as odd is that most of the xfer_hdr is for moving data from peer to peer, and this is really specific to the unexpected message path.

Reply:
I'm ok with adding an extra field if it doesn't impact performance - I'll defer to Seth for that decision.
It is specific to the unexpected message path but the next is specific to having a message actively in the queue. It's just a different path - ie the "next" field is only used when it is in the queue, the "ep" field is only used when it is not in the queue. The field itself just holds a piece of information for who is tracking/using the entry.

Reply:
No more fields in the header lol!

Reply:
If we did want to make this little change into a big change... we should add a `void* ep` (or a `struct util_ep *ep`) here:
```
struct fi_ops_srx_owner {
	size_t	size;
	int	(*get_msg)(struct fid_peer_srx *srx, fi_addr_t addr,
			size_t size, struct fi_peer_rx_entry **entry);
	int	(*get_tag)(struct fid_peer_srx *srx, fi_addr_t addr,
			size_t size, uint64_t tag, struct fi_peer_rx_entry **entry);
	int	(*queue_msg)(struct fi_peer_rx_entry *entry);
	int	(*queue_tag)(struct fi_peer_rx_entry *entry);
	void	(*foreach_unspec_addr)(struct fid_peer_srx *srx,
			fi_addr_t (*get_addr)(struct fi_peer_rx_entry *));

	void	(*free_entry)(struct fi_peer_rx_entry *entry);
};

struct fi_ops_srx_peer {
	size_t	size;
	int	(*start_msg)(struct fi_peer_rx_entry *entry);
	int	(*start_tag)(struct fi_peer_rx_entry *entry);
	int	(*discard_msg)(struct fi_peer_rx_entry *entry);
	int	(*discard_tag)(struct fi_peer_rx_entry *entry);
};
```

Reply:
the problem is that we already have a void * in the entry to specify peer-specific information...

Reply:
This diff hurts our 96rank osu_alltoallw from 154us -> 162us... we either need to accept this change, or modify the peer API.
```
└─╼ git diff
diff --git a/prov/sm2/src/sm2.h b/prov/sm2/src/sm2.h
index 8e136517..6905ab05 100644
--- a/prov/sm2/src/sm2.h
+++ b/prov/sm2/src/sm2.h
@@ -108,14 +108,12 @@ enum {
  *     user_data - the message
  */
 struct sm2_xfer_hdr {
-       union {
-               volatile long int next;
-               struct sm2_ep *ep;
-       };
+       volatile long int next;
        uint64_t size;
        uint64_t cq_data;
        uint64_t tag;
        uint64_t context;
+       struct sm2_ep *ep;
        uint32_t op;
        uint32_t op_flags;
        uint32_t proto;
```

Reply:
I can accept the PR as is, I'm fine with the union for next and *ep.

That said, I'm scratching my head trying to understand how it could possibly affect performance.  My mental model is not tracking.